### ISSUE TITLE
Parameter lookup needs to be optimized

### ISSUE THREAD

**Issue by j-wags** [openforcefield/openforcefield#383]
**Describe the problem**
This became a serious issue for our initial parameterization sprint. @yudongqiu made a timely workaround so this isn't immediately necessary, but we need to revisit this relatively urgently. If we don't, refactoring our internal parameter structure could break our overall fitting workflow.

https://github.com/leeping/forcebalance/pull/149#issue-295570493

**Comment by davidlmobley** [comment]
@j-wags you may have missed this issue, which will dramatically accelerate parameter assignment with almost no work: https://github.com/openforcefield/openforcefield/issues/51

Basically, we currently process in a "last one wins" manner which means many many substructure searches need to be done to assign any parameter. Reversing the hierarchy and then assigning in a "first one wins" manner (or to put it another way, processing bottom-up) will give much better performance but be functionally identical.

**Comment by yudongqiu** [comment]
I think it might be good to clarify it a little bit here.

The update leeping/forcebalance#149 is focusing on updating parameters in created `ForceField` objects. This is not related to the assigning of parameters to molecule topologies, which happens `forcefield.create_openmm_system` is called.

The updates in leeping/forcebalance#149 simply accelerates the lookup for the internal parameter representation in the `ForceField` object that allows changing the values instantly.

The "assigning of parameters to molecule topologies" procedure is also accelerated, in my implementation of `smirnoff_hack.py` in ForceBalance.

**Comment by j-wags** [comment]
Yes @yudongqiu is correct. Apologies, I was in a rush and wasn't clear.

The act of retrieving a specific `ParameterHandler`, looking up one of its parameters via its SMIRKS, and finding one attribute of that SMIRKS is the slow process here. I think this is due to the fact that a `ParameterList` inherits from `list`, but we support looking parameters up in it using `ParameterList[smirks]`, which makes a reasonable person think it's a dict, but it's not. Structurally, we can't make it a dict, because of the reasons [here](https://github.com/openforcefield/openforcefield/blob/master/openforcefield/typing/engines/smirnoff/parameters.py#L671-L674). Though, with the new `ValidatedList` class, we might be able to safely maintain a cached lookup dict. We should revisit this once we're through the initial FF crunch.

#50 is still a great idea and a simple fix, and we should implement it once we have some time for cleanups.

**Comment by andrrizzi** [comment]
> Though, with the new `ValidatedList` class, we might be able to safely maintain a cached lookup `dict`.

I think that's a great idea. It shouldn't be difficult to make the changes.

**Comment by andrrizzi** [comment]
Or maybe implementing a similar `ValidatedDict` could be the way.
